finality-aleph/src/justification/compatibility.rs

```diff
@@ -27,22 +32,242 @@ impl From<AlephJustificationV1> for AlephJustification {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub enum JustificationDecoding {
+enum VersionedAlephJustification {
+    // Most likely from the future.
+    Other(u32, Vec<u8>),
     V1(AlephJustificationV1),
     V2(AlephJustification),
-    Err,
 }
 
-pub fn backwards_compatible_decode(justification_raw: Vec<u8>) -> JustificationDecoding {
+impl Encode for VersionedAlephJustification {
+    fn size_hint(&self) -> usize {
+        use VersionedAlephJustification::*;
+        let version_and_bytes_size = 2 * size_of::<u32>();
+        version_and_bytes_size
+            + match self {
+                Other(_, payload) => payload.size_hint(),
+                V1(justification) => justification.size_hint(),
+                V2(justification) => justification.size_hint(),
+            }
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        use VersionedAlephJustification::*;
+        match self {
+            Other(version, payload) => {
+                let mut result = version.encode();
+                let mut payload = payload.encode();
+                let num_bytes = payload.len() as u32;
+                result.append(&mut num_bytes.encode());
+                result.append(&mut payload);
+                result
+            }
+            V1(justification) => {
+                let mut result = (1_u32).encode();
+                let mut justification = justification.encode();
+                let num_bytes = justification.len() as u32;
+                result.append(&mut num_bytes.encode());
+                result.append(&mut justification);
+                result
+            }
+            V2(justification) => {
+                let mut result = (2_u32).encode();
+                let mut justification = justification.encode();
+                let num_bytes = justification.len() as u32;
+                result.append(&mut num_bytes.encode());
+                result.append(&mut justification);
+                result
+            }
```

Comment:
nit: all arms could be handled by one function `fn smt<E: Encode>(version, payload: E) -> Vec<u8>`

Reply:
the version could be of the `type Version = u16` to enforce always one type for it

Reply:
Slightly different than you suggested so that it works for `Other` as well, but done.

Also possibly fixed a bug on the way and added a test for it (in wrong order unfortunately, so I'm not quite sure the bug was there).